src/resource.rs

```diff
@@ -2,35 +2,325 @@
+        match self.signature_func {
+            SignatureFunc::CustomFunc(_) => true,
+            SignatureFunc::FromYAML { .. } => false,
+        }
+    }
 
-use crate::ops::{OpDef, OpaqueOp};
-use crate::types::custom::CustomType;
+    /// Fallibly returns a Hugr that may replace an instance of this OpDef
+    /// given a set of available resources that may be used in the Hugr.
+    pub fn try_lower(&self, args: &[TypeArg], available_resources: &ResourceSet) -> Option<Hugr> {
+        match &self.lower_func {
+            LowerFunc::None => None,
+            LowerFunc::FixedHugr(req_res, h) => {
+                if available_resources.is_superset(req_res) {
+                    Some(h.clone())
+                } else {
+                    None
+                }
+            }
+            LowerFunc::CustomFunc(f) => {
+                f.try_lower(&self.name, args, &self.misc, available_resources)
+            }
+        }
+    }
+}
+
+/// A declaration of an opaque type.
+/// Note this does not provide any way to create instances
+/// - typically these are operations also provided by the Resource.
+#[derive(Debug, serde::Serialize, serde::Deserialize)]
+pub struct TypeDef {
+    /// The unique name of the type
+    pub name: SmolStr,
+    /// Declaration of type parameters. The TypeDef must be instantiated
+    /// with the same number of [`TypeArg`]'s to make an actual type.
+    ///
+    /// [`TypeArg`]: crate::types::type_param::TypeArg
+    pub args: Vec<TypeParam>,
+}
 
 /// A unique identifier for a resource.
 ///
 /// The actual [`Resource`] is stored externally.
 pub type ResourceId = SmolStr;
 
 /// A resource is a set of capabilities required to execute a graph.
-#[derive(Clone, Debug, Default, serde::Serialize, serde::Deserialize)]
+#[derive(Debug, Default, serde::Serialize, serde::Deserialize)]
 pub struct Resource {
     /// Unique identifier for the resource.
     pub name: ResourceId,
-    /// Set of resource dependencies required by this resource.
+    /// Other resources defining types used by this resource.
+    /// That is, an upper-bound on the types that can be returned by
+    /// computing the signature of any operation in this resource,
+    /// for any possible [TypeArg].
     pub resource_reqs: ResourceSet,
     /// Types defined by this resource.
-    pub types: Vec<CustomType>,
+    types: HashMap<SmolStr, TypeDef>,
     /// Operation declarations with serializable definitions.
-    pub operations: Vec<OpDef>,
-    /// Opaque operation declarations that do not expose their definitions.
-    pub opaque_operations: Vec<OpaqueOp>,
+    #[serde(serialize_with = "elide_arcs", deserialize_with = "reinstate_arcs")]
+    operations: HashMap<SmolStr, Arc<OpDef>>,
+}
+
+fn elide_arcs<S: Serializer>(
+    _ops: &HashMap<SmolStr, Arc<OpDef>>,
+    _serializer: S,
+) -> Result<S::Ok, S::Error> {
+    // serde doesn't like Arc - write it out as if it were a HashMap<SmolStr, OpDef>
+    todo!()
```

Comment:
Well, they do affect serialization of *resources*, but not of *ops* or Hugrs; we expect to pass the serialized YAML around with the Hugr, but that is very much a TODO at the moment. I'm not sure we necessarily need a non-YAML Resource format, or whether we do any serialization of Resources atm. @mark-koch perhaps?

If we do need this now, this may have to wait awhile, I've not figured out how to make serde comply yet...